src/core/core/network.cpp

```diff
@@ -398,15 +398,31 @@ namespace dsn
         return ret;
     }
 
-    void rpc_session::on_recv_message(message_ex* msg, int delay_ms)
+    bool rpc_session::on_recv_message(message_ex* msg, int delay_ms)
     {
         msg->to_address = _net.address();
         msg->io_session = this;
 
         if (msg->header->context.u.is_request)
         {
-            dbg_dassert(!is_client(), 
-                "only rpc server session can recv rpc requests");
+            // ATTENTION: need to check if self connection occurred.
+            //
+            // When we try to connect some socket in the same host, if we don't bind the client to a specific port,
+            // operating system will provide ephemeral port for us. If it's happened to be the one we want to connect to,
+            // it causes self connection.
+            //
+            // The case is:
+            // - this session is a client session
+            // - the remote address is in the same host
+            // - the remote address is not listened, which means the remote port is not occupied
+            // - operating system chooses the remote port as client's ephemeral port
+            if (is_client() && msg->header->from_address == _net.engine()->primary_address())
```

Comment:
msg->to_address may be not the same as primary_address, but msg->header->from_address is always set as primary address, as in "rpc_engine::call()".